Improve changelog generation script

behackl opened the pull request:
<!--
Thanks for your contribution to ManimCommunity!

Before filling in the details, ensure:
- Your local changes are up-to-date with ManimCommunity/manim
  
- The title of your PR gives a descriptive summary to end-users. Some examples:
  - Fixed last animations not running to completion
  - Added gradient support and documentation for SVG files
  Examples of what *NOT* to do:
  - "fixed that styling issue" - not descriptive enough
  - "fixed issue #XYZ" - end-user needs to do further research
-->
## Changelog / Overview

- Added some new flags to the changelog generation script (`-t` for specifying the new release name, `-a` for including additional PRs),
- adapted the script documentation accordingly,
- removed the overlooked heading from `changelog.rst`,
- and added a proper title + date to the v0.5.0 changelog. (Too late, but whatever...)

<!-- Optional (Recommended): a detailed overview of the PR for the upcoming
release's changelog entry. Useful for when the PR title isn't enough. 

DO NOT REMOVE THE FOLLOWING CHANGELOG LINES, EVEN IF YOU DON'T USE THEM.-->
<!--changelog-start-->
<!--changelog-end-->

## Further Comments
<!-- Optional: any further comments that might be useful for reviewers. -->

## Checklist
- [x] I have read the [Contributing Guidelines](https://docs.manim.community/en/latest/contributing.html)
- [x] I have written a descriptive PR title (see top of PR template for examples)
- [x] I have written a changelog entry for the PR or deem it unnecessary
- [x] My new functions/classes either have a docstring or are private
- [x] My new functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [x] My new documentation builds, looks correctly formatted, and adds no additional build warnings
<!-- Once again, thanks for contributing to ManimCommunity! -->


<!-- Do not modify the lines below. These are for the reviewers of your PR -->
## Reviewer Checklist
- [ ] The PR title is descriptive enough
- [ ] The PR is labeled correctly
- [ ] The changelog entry is completed if necessary
- [ ] Newly added functions/classes either have a docstring or are private
- [ ] Newly added functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [ ] Newly added documentation builds, looks correctly formatted, and adds no additional build warnings

behackl commented:
Some further things I've noticed (that should also be resolved and/or investigated):

- We should probably have a different label for bugfixes, the heading "bug" is somewhat irritating,
- The PR reverting the merge of mobject and OpenGLMobject is not included in the changelog.

jsonvillanueva commented:
Woah, I was [just about to make a PR](https://discord.com/channels/581738731934056449/581738732646957057/827448199832141874) for this but let's use this instead as you beat me to it 😄

jsonvillanueva commented:
>     * The PR reverting the merge of mobject and OpenGLMobject is not included in the changelog.

That's because sort_pr_by_labels doesn't take into account the "revert" label

jsonvillanueva commented on docs/source/changelog.rst:
```diff
@@ -2,11 +2,6 @@
 
-****************
-Upcoming release
-****************
```
I sorta like this section, but it's not really used unfortunately. It'd also only make sense for this to be used during latest so I'm fine with this being removed

jsonvillanueva commented on docs/source/changelog/0.5.0-changelog.rst:
```diff
@@ -1,3 +1,9 @@
+******
+v0.5.0
+******
+
+:Date: April 1, 2021
+
```
Since this is being updated, can you run your version of the script and use its output for the file as the new changelog? At least this way the latest documentation will be correct for the contributors of the release.

jsonvillanueva commented on scripts/dev_changelog.py:
```diff
@@ -9,24 +10,31 @@
 From the bash command line with $GITHUB token::
-    $ ./scripts/dev_changelog.py $GITHUB v0.3.0..v0.4.0 -o 0.4.0-changelog.rst
+
+    $ ./scripts/dev_changelog.py $GITHUB v0.3.0..HEAD -t v0.4.0 -o 0.4.0-changelog.rst
```
I like this! I can update the GitHub Wiki accordingly.

One thing I'll add is that `..HEAD` portion of `v0.3.0..HEAD` seems unnecessary with this. I think the revision_range logic can be modified

behackl replied:
Agreed. I'll have to leave for a while, feel free to simply push to my branch if you get to refactor it in the meantime.

behackl commented:
Sorry! I should have mentioned that I'm also looking into it; I'd be happy to switch to your branch in case your changes are more stable than mine are. 😅 

> That's because sort_pr_by_labels doesn't take into account the "revert" label

Then we should add the label to the list I guess.

> What would you suggest? I think the label bug should be different from the pull request label bug, just to diversify make the meaning make sense

Bugfix? Alternatively, we could also introduce a dictionary where labels are mapped to section titles; that would avoid the singular/plural problem.

jsonvillanueva commented on scripts/dev_changelog.py:
```diff
@@ -250,5 +273,11 @@ def main(token, revision_range, outfile=None):
+        "-t", "--tag", type=str, help="the tag of the new release"
+    )
+    parser.add_argument(
+        "-a", "--additional", type=int, action="append", help="include an additional PR in the changelog that has not been recognized automatically. flag can be added more than once."
```
I like the idea of this flag, though it might make more sense to use click :P
I know for certain there is a way to take in as many PR nums with just a single instance of -a as say a , separated list rather easily

jsonvillanueva commented:
> Sorry! I should have mentioned that I'm also looking into it; I'd be happy to switch to your branch in case your changes are more stable than mine are. 😅

Nah, it's more/less the same. I like the new options you introduced

> Then we should add the label to the list I guess.

https://github.com/ManimCommunity/manim/pull/1231/files#diff-c461d965355498ee3bd152eda8723432a8d81b2bf09246f0cc8bf38c96d406a4R114-R140 There's definitely a better way to do this. I just did the first thing that came to mind... really it'd be nice to take into account all of the PR labels but I'm not sure how that would translate into the changelog-start/changelog-end logic...

behackl commented:
I've simplified the logic behind sorting into labels by a bit and updated the changelog by the corresponding output. I'd like to have this PR merged rather sooner than later (in order to make the release announcements). Would you be okay with moving the improvements for the CLI args to a follow-up PR?

behackl commented:
Oh, and I forgot to mention: we can now have more sophisticated section titles in the changelog; I've proposed a few (and slightly reordered / condensed the existing ones).

jsonvillanueva commented:
I've got some Click changes to the CI that I can maybe merge in the next 30 mins if the announcement is urgent -- but otherwise I can make a separate PR for this.

behackl commented:
> I've got some Click changes to the CLI that I can maybe merge in the next 30 mins if the announcement isn't urgent -- but otherwise I can make a separate PR for this.

I'll simply keep this open while I'm looking for what causes the RTD problem. If I find the problem first and am able to fix it, then please add further changes in a follow-up -- otherwise feel free to push your changes here. 👍

behackl commented:
Let's move this along so that we can make the announcement.
